### PR TITLE
Add isCompressed to PutRequest V5

### PR DESCRIPTION
**Pull request by panchow**
A new version of PutRequest V5 is introduced but disabled, so it continues to use V4 as the default.
The new version has aded the "isCompressed" field, serialization and de-serialization would apply PutRequest V5 only if the version is set to V5.

Added unit test to verify behavior.

### CODE REVIEW

File: ambry-protocol/src/test/java/com/github/ambry/protocol/PutRequestTest.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import org.junit.Assert;
+import org.junit.Test;
+
+
+/**
+ * PutRequest testing.  Add more unit tests here to test all public methods and code changes to the PutRequest.
+ */
+public class PutRequestTest {
```
Comment (justinlin-linkedin) [inline ambry-protocol/src/test/java/com/github/ambry/protocol/PutRequestTest.java:34]: we have a test class RequestResponseTest where we do all the Request/Response test, please move this test case to it.

**Review by justinlin-linkedin — COMMENTED** [review]
The rest looks good to me, just one minor comments

File: ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java [hunk 2]
```diff
@@ -44,6 +44,7 @@ public class PutRequest extends RequestOrResponse {
   protected final BlobProperties properties;
   protected final BlobType blobType;
   protected final ByteBuffer blobEncryptionKey;
+  protected boolean isCompressed;
```
Comment (snalli) [inline ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java:47]: final

File: ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java [hunk 3]
```diff
@@ -213,6 +268,9 @@ protected void prepareBuffer() {
     if (keyLength > 0) {
       bufferToSend.writeBytes(blobEncryptionKey);
     }
+    if (versionId >= PUT_REQUEST_VERSION_V5) {
```
Comment (snalli) [inline ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java:271]:
```
if (versionId >= PUT_REQUEST_VERSION_V5 && isCompressed) {
 bufferToSend.writeByte(1);
}
```

File: ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java [hunk 4]
```diff
@@ -213,6 +268,9 @@ protected void prepareBuffer() {
...
+      bufferToSend.writeByte(isCompressed ? 1 : 0);
```
Comment (snalli) [inline ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java:272]:
```
    if (versionId >= PUT_REQUEST_VERSION_V5) {
      bufferToSend.writeByte(Boolean.compare(isCompressed, true));
    }
```

**Comment by snalli** [comment]
Added unit to verify behavior -> Added unit test to verify behavior.

File: ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java (hunk 4 above)
Comment (panchow) [reply ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java:272]: Isn't it easier to read using (isCompressed ? 1 : 0)? This pattern is used in the rest of the Ambry code too.

File: ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java (hunk 2 above)
Comment (panchow) [reply ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java:47]: thank you

File: ambry-protocol/src/test/java/com/github/ambry/protocol/PutRequestTest.java (hunk 1 above)
Comment (panchow) [reply ambry-protocol/src/test/java/com/github/ambry/protocol/PutRequestTest.java]: Moved the test to the RequestResponseTest class.

**Review by justinlin-linkedin — APPROVED** [review]
lgtm

File: ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java (hunk 4 above)
Comment (snalli) [reply ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java:272]: Both are easy enough but if Java provides a method, I think it's better to use it. 
Agreed this pattern is in rest of Ambry, but we can try changing that henceforth.

Comment (snalli) [reply ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java:272]: But nvm, the pr lgtm.

**Pull request merged**